src/mongo/Config.class.php

```diff
@@ -1759,7 +1760,23 @@ protected function getConnectionForDataSource($dataSource)
         }
         if(!isset($this->connections[$dataSource]))
         {
-            $this->connections[$dataSource] = new \MongoClient($ds['connection'], $connectionOptions);
+            $retries = 1;
+            $exception = null;
+
+            do {
+                try {
+                    $this->connections[$dataSource] = new \MongoClient($ds['connection'], $connectionOptions);
+                } catch (\MongoConnectionException $e) {
```

Comment:
We need to specifically look for an error code here on the exception. If we had a situation where the cluster was offline, then it's going to lock up Apache processes for 30 seconds. This is going to lock up the servers and cause untold woe.

Better to be specific about the exception and try and capture a step down.